### ISSUE TITLE
NIO just updated their package manifest to 5.6 and it started erroring when building the tests 

### ISSUE THREAD

**Issue by buttaface** [apple/swift-package-manager#5717]
### Description

@tomerd made the change in apple/swift-nio#2235 and that now causes an error on linux and Android when compiling `LinuxMain.swift`. Workarounds are to revert that pull or add `--enable-test-discovery`, which avoids compiling `LinuxMain.swift`.

### Expected behavior

The NIO tests to build and run.

### Actual behavior

This compilation error:
```
/home/butta/swift-nio/Tests/LinuxMain.swift:146:1: error: expressions are not allowed at the top level
(LinuxMainRunnerImpl() as LinuxMainRunner).run()
```

### Steps to reproduce

1. Check out swift-nio.
2. Build the tests with `swift test`.

### Swift Package Manager version/commit hash

5.6.2 and latest 5.7 and trunk snapshots

### Swift & OS version (output of `swift --version && uname -a`)

Swift 5.6.2/5.7/5.8 running on Ubuntu 20.04 x86_64 or Android 12 AArch64

**Comment by neonichu** [comment]
That's expected since we were assuming that any non-main.swift file is always using @main when using a 5.6+ tools-versions -- https://github.com/apple/swift-package-manager/pull/5688 improves this behavior already, so I don't think there is more to do for SwiftPM here.

**Comment by tomerd** [comment]
https://github.com/apple/swift-package-manager/pull/5688 should fix it on 5.7 and main

**Comment by buttaface** [comment]
Alright, so people using 5.6 who set `swift-tools-version` to 5.6 are expected to patch that NIO manifest like this?
```
diff --git a/Tests/LinuxMain.swift b/Tests/LinuxMain.swift
index eacd4b02..a6a92e48 100644
--- a/Tests/LinuxMain.swift
+++ b/Tests/LinuxMain.swift
@@ -40,10 +40,9 @@ import XCTest
 // without the compiler noticing that we're calling a deprecated function.
 // This hack exists so we can deprecate individual tests which test deprecated functionality without
 // getting a compiler warning...
-protocol LinuxMainRunner { func run() }
-class LinuxMainRunnerImpl: LinuxMainRunner {
+@main class LinuxMainRunnerImpl {
    @available(*, deprecated, message: "not actually deprecated. Just deprecated to allow deprecated tests (which test deprecated functionality) without warnings")
-   func run() {
+   static func main() {
        XCTMain([
              testCase(AcceptBackoffHandlerTest.allTests),
              testCase(AdaptiveRecvByteBufferAllocatorTest.allTests),
@@ -144,5 +143,4 @@ class LinuxMainRunnerImpl: LinuxMainRunner {
         ])
     }
 }
-(LinuxMainRunnerImpl() as LinuxMainRunner).run()
 #endif
```
That patch works with both 5.6.2 and the latest Aug. 9 trunk snapshot build for linux, but I think you should probably get #5687 into a 5.6 patch release, or this `@main` requirement is going to break many CIs that test older Swift versions such as 5.6.

As I just noted on that NIO pull, #5687 does fix this issue, but 5.6 is currently not patched with it.

**Comment by tomerd** [comment]
this problem was introduced in 5.5 actually, and 5.6 also suffers from it. SwiftNIO adopting 5.6 brings to to bear because it also using `LinuxMain.swift`, which is less common at this point. I think the workaround for folks that run swift-nio tests  is to pass `--enable-test-discovery` not change `LinuxMain.swift`

**Comment by buttaface** [comment]
> this problem was introduced in 5.5 actually, and 5.6 also suffers from it.

You're right, I just tried it with Swift 5.5.3 and it errors the same.

> SwiftNIO still supports 5.4 and that does not have @main

I think it does, but don't know if this patch would work with it.

5.5 having this issue too definitely makes fixing 5.6 less important, but would be nice to get #5687 into a 5.6 patch release. If that's too late because of the impending 5.7, I understand.

**Comment by buttaface** [comment]
Hmm, the fix was already reverted on 5.7.0, #5684, hope it makes it into 5.7.1.
